applications/machine_learning/src/modules/ml_runner.c

```diff
@@ -131,6 +124,31 @@ static void start_prediction(void)
 	}
 }
 
+static void result_ready_cb(int err)
+{
+	k_sched_lock();
+
+	bool drop_result = (err) || (ml_control & ML_DROP_RESULT) || (state == STATE_ERROR);
+
+	if (err) {
+		LOG_ERR("Result ready callback returned error (err: %d)", err);
+		report_error();
+	} else {
+		ml_control &= ~ML_DROP_RESULT;
+		ml_control &= ~ML_RUNNING;
+
+		if (state == STATE_ACTIVE) {
+			start_prediction();
```

Comment:
does it have to be under a lock? I think it would be enough to check state and start after lock is released.
in the end your only gain will be to make prediction not started in very specific race case (other races can still happen) and the cost is to keep the lock longer

Reply:
The `start_prediction` call starts the prediction and updates machine learning state tracked by the module. At least the state update should be done under lock. It also cleans the buffer if needed (the function was designed to simplify suspending and resuming ml_runner).
To keep state management simpler, I prefer not to split this functionality. Also the overhead is not big, since function starting prediction is not very complex and we are not blocking interrupts here.

Reply:
Easy way to suspend/resume the module could be also useful after we introduce sensor triggers.

Reply:
discussed offline